src/test/cls_rbd/test_cls_rbd.cc

```diff
@@ -2317,6 +2327,417 @@ TEST_F(TestClsRbd, mirror_snapshot) {
+
+  cls::rbd::MirrorGroup read_group;
+  ASSERT_EQ(0, mirror_group_get(&ioctx, "group_id1", &read_group));
+  ASSERT_EQ(read_group, group1);
+  ASSERT_EQ(0, mirror_group_get(&ioctx, "group_id2", &read_group));
+  ASSERT_EQ(read_group, group2);
+  ASSERT_EQ(0, mirror_group_get(&ioctx, "group_id3", &read_group));
+  ASSERT_EQ(read_group, group3);
+
+  std::map<std::string, cls::rbd::MirrorGroup> expected_groups;
+  ASSERT_EQ(0, mirror_group_list(&ioctx, "", 1, &groups));
+  expected_groups = {{"group_id1", group1}};
+  ASSERT_EQ(expected_groups, groups);
+
+  ASSERT_EQ(0, mirror_group_list(&ioctx, "group_id1", 2, &groups));
+  expected_groups = {{"group_id2", group2}, {"group_id3", group3}};
+  ASSERT_EQ(expected_groups, groups);
+
+  ASSERT_EQ(0, mirror_group_remove(&ioctx, "group_id2"));
+  ASSERT_EQ(-ENOENT, mirror_group_get_group_id(&ioctx, "uuid2", &group_id));
+  ASSERT_EQ(-EBUSY, mirror_group_remove(&ioctx, "group_id1"));
+
+  ASSERT_EQ(0, mirror_group_list(&ioctx, "", 3, &groups));
+  expected_groups = {{"group_id1", group1}, {"group_id3", group3}};
+  ASSERT_EQ(expected_groups, groups);
+
+  group1.state = cls::rbd::MIRROR_GROUP_STATE_DISABLING;
+  group3.state = cls::rbd::MIRROR_GROUP_STATE_DISABLING;
+  ASSERT_EQ(0, mirror_group_set(&ioctx, "group_id1", group1));
+  ASSERT_EQ(0, mirror_group_get(&ioctx, "group_id1", &read_group));
+  ASSERT_EQ(read_group, group1);
+  ASSERT_EQ(0, mirror_group_set(&ioctx, "group_id3", group3));
+  ASSERT_EQ(0, mirror_group_remove(&ioctx, "group_id1"));
+  ASSERT_EQ(0, mirror_group_remove(&ioctx, "group_id3"));
+
+  ASSERT_EQ(0, mirror_group_list(&ioctx, "", 3, &groups));
+  expected_groups = {};
+  ASSERT_EQ(expected_groups, groups);
+}
+
+TEST_F(TestClsRbd, mirror_group_status) {
+  struct WatchCtx : public librados::WatchCtx2 {
+    librados::IoCtx *m_ioctx;
+
+    explicit WatchCtx(librados::IoCtx *ioctx) : m_ioctx(ioctx) {}
+    void handle_notify(uint64_t notify_id, uint64_t cookie,
+                       uint64_t notifier_id, bufferlist& bl_) override {
+      bufferlist bl;
+      m_ioctx->notify_ack(RBD_MIRRORING, notify_id, cookie, bl);
+    }
+    void handle_error(uint64_t cookie, int err) override {}
+  };
+
+  map<std::string, cls::rbd::MirrorGroup> groups;
+  map<std::string, cls::rbd::MirrorGroupStatus> statuses;
+  std::map<cls::rbd::MirrorGroupStatusState, int32_t> states;
+  std::map<std::string, entity_inst_t> instances;
+  cls::rbd::MirrorGroupStatus read_status;
+  entity_inst_t read_instance;
+  uint64_t watch_handle;
+  librados::IoCtx ioctx;
+
+  ASSERT_EQ(0, _rados.ioctx_create(_pool_name.c_str(), ioctx));
+  ioctx.remove(RBD_MIRRORING);
+
+  int64_t instance_id = librados::Rados(ioctx).get_instance_id();
+
+  // Test list fails on nonexistent RBD_MIRRORING object
+
+  ASSERT_EQ(-ENOENT, mirror_group_status_list(&ioctx, "", 1024, &groups,
+	  &statuses));
+
+  // Test status set
+
+  cls::rbd::MirrorGroup group1("uuid1", cls::rbd::MIRROR_IMAGE_MODE_JOURNAL,
```

Comment:
Should the cls functions prevent setting the mode to `MIRROR_IMAGE_MODE_JOURNAL` for a group?